### PR TITLE
feat: Added custom label and annotations to the pods of the server, repo, application set controller, and application controller

### LINKED ISSUE #1530: Support for custom labels and annotations to Server, repo, application set and application controller
**Is your feature request related to a problem? Please describe.**
Hi! I need a feature that allows the addition of custom labels and annotations to pods deployed by the operator, specifically in the server, repos, application set and application controller components. This is required to enable direct metric scraping within the pod using Vector in my environment.

**Describe the solution you'd like**
Something like that:
```
spec:
  server:
    customPodLabels:
      custom: label
      asdf: server
    customPodAnnotations:
      asdf: server
      custom: annotation
  repo:
    customPodLabels:
      custom: label
      asdf: repo
    customPodAnnotations:
      asdf: repo
      custom: annotation
  controller:
    customPodLabels:
      custom: label
      asdf: controller
    customPodAnnotations:
      asdf: controller
      custom: annotation
  applicationSet:
    customPodLabels:
      custom: label
      asdf: applicationSet
    customPodAnnotations:
      asdf: applicationSet
      custom: annotation
```
**Additional context**
When I saw that the operator doesn´t support this feature, I found [this issue](https://github.com/argoproj-labs/argocd-operator/issues/637) and [this PR](https://github.com/argoproj-labs/argocd-operator/pull/1009).  Inspired by the code that @AlexanderThaller did in his branch, I started working on my own version, i have [a branch](https://github.com/enriqueav99/argocd-operator/tree/custom-labels-annotations) tfor testing and I accidentally opened [a PR](https://github.com/argoproj-labs/argocd-operator/pull/1529). I want to make another branch with diferent commits for a better review.

### PR DESCRIPTION
**Pull request by enriqueav99**
**What type of PR is this?**

> /kind enhancement

**What does this PR do / why we need it**:

Add the feature to add custom labels and annotations  to the pods of the server, repo, application set controller, and application controller.

**Have you updated the necessary documentation?**

* [x] Documentation update is required by this PR.
* [x] Documentation has been updated.

**Which issue(s) this PR fixes**:

Fixes #1530 #637

**How to test changes / Special notes to the reviewer**:
Apply the following yaml
```
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: argocd-sample
spec:
  server:
    customPodLabels:
      custom: label
      custom2: server
    customPodAnnotations:
      custom: annotation
      custom2: server
  repo:
    customPodLabels:
      custom: label
      custom2: repo
    customPodAnnotations:
      custom: annotation
      custom2: repo
  controller:
    customPodLabels:
      custom: label
      custom2: controller
    customPodAnnotations:
      custom: annotation
      custom2: controller
  applicationSet:
    customPodLabels:
      custom: label
      custom2: applicationSet
    customPodAnnotations:
      custom: annotation
      custom2: applicationSet
```

### CODE REVIEW

**Comment by enriqueav99** [comment]
I deleted the test I made because I found it unnecessary and it wasn't working well in the CI. However, it passes when I run it locally (A classic)

**Comment by enriqueav99** [comment]
I don't understand why the _kuttl/harness/1-036_validate_role_rolebinding_for_source_namespace_ test sometimes passes and sometimes doesn't, even though I haven't changed any code. For example, in the last commit, I deleted a test I created, and it didn't pass, but in the previous two commits, it did pass.

**Comment by svghadi** [comment]
Hi @enriqueav99,
> kuttl/harness/1-036_validate_role_rolebinding_for_source_namespace

is a known flaky test. You can ignore it.

**Comment by enriqueavindi** [comment]
> Hi @enriqueav99,
> 
> > kuttl/harness/1-036_validate_role_rolebinding_for_source_namespace
> 
> is a known flaky test. You can ignore it.

Thanks for your reply! If it is a known "issue" I think the PR is ready to be reviewed

**Comment by enriqueav99** [comment]
> Hi @enriqueav99,
> 
> > kuttl/harness/1-036_validate_role_rolebinding_for_source_namespace
> 
> is a known flaky test. You can ignore it.
Thaks for your reply! If it is a knwon "issue" I think taht the PR is ready to be reviewed

**Comment by enriqueav99** [comment]
Hi! Could someone please review this? I think it's something simple and it would be really helpful for me at work 😄

**Comment by svghadi** [comment]
Hi @enriqueav99, thanks for the PR. I will try to get to it in next few days.

**Comment by svghadi** [comment]
Hi @enriqueav99 , can we the naming convention from 
- `customPodLabels` to `labels`
- `customPodAnnotations ` to `annotations`

**Comment by enriqueav99** [comment]
> Hi @enriqueav99 , can we the naming convention from 
> - `customPodLabels` to `labels`
> - `customPodAnnotations ` to `annotations`

Hi! I changed the name of them! Thanks for yor time, any other feedback or review Will be well recived :D

File: tests/k8s/1-043_validate_custom_labels_annotations/01-install.yaml
Comment (svghadi) [inline tests/k8s/1-043_validate_custom_labels_annotations/01-install.yaml:1]: Please add the `01-assert.yaml` file along with this one, which should contain the YAML representation of the expected results.

Example:
```
# 01-assert.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: argocd-sample-server
spec:
  template:
    metadata:
      annotations:
        custom: annotation
        custom2: server
      labels:
        app.kubernetes.io/name: argocd-sample-server
        custom: label
        custom2: server
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: argocd-sample-repo-server
spec:
  template:
    metadata:
      annotations:
        custom: annotation
        custom2: repo
      labels:
        app.kubernetes.io/name: argocd-sample-repo-server
        custom: label
        custom2: repo
``` 

Can you also cover case where labels & annotations are removed from `ArgoCD` CR?

File: docs/usage/custom_labels_annotations.md
Comment (svghadi) [inline docs/usage/custom_labels_annotations.md:1]: Add a link to this new file here: https://github.com/argoproj-labs/argocd-operator/blob/603addacfb9d4c00e33d13bc561374b8a041366b/mkdocs.yml#L61 
After this line.

File: deploy/olm-catalog/argocd-operator/0.12.0/argoproj.io_argocds.yaml
Comment (svghadi) [inline deploy/olm-catalog/argocd-operator/0.12.0/argoproj.io_argocds.yaml:1]: Can we revert this changes? The CSV should be updated only for latest version i.e 0.13.0.

**Review by svghadi — COMMENTED** [review]
Thanks @enriqueav99. 
Ran some tests, overall looks good. Left some suggestions.

File: docs/usage/custom_labels_annotations.md
Comment (enriqueav99) [reply docs/usage/custom_labels_annotations.md:1]: Added the link

File: deploy/olm-catalog/argocd-operator/0.12.0/argoproj.io_argocds.yaml
Comment (enriqueav99) [reply deploy/olm-catalog/argocd-operator/0.12.0/argoproj.io_argocds.yaml:1]: reverted!

**Comment by enriqueav99** [comment]
Thanks for the review @svghadi! Please, if another change is needed, let me know 😄

**Review by svghadi — COMMENTED** [review]
LGTM. Thanks @enriqueav99 for the contribution! Once you resolve the conflict, the PR will be good to go.

**Comment by enriqueavindi** [comment]
All conflicts are solved, thanks @svghadi for your help!

**Comment by enriqueav99** [comment]
All conflicts are solved, thanks @svghadi for your help!

**Review by svghadi — APPROVED** [review]
Awesome 🎉 . Thanks a lot @enriqueav99 !!

**Comment by enriqueav99** [comment]
Thanks @svghadi very much 😄 . Do you know if an early release is planned? I need this urgently for my organization, and we want to create a schedule to make the necessary changes to include this new feature and start monitoring our ArgoCD clusters 🚀

**Comment by svghadi** [comment]
We do one release per quarter. The next planned release is somewhere at start of December. I am afraid that an early release won't be possible. Currently we do manual releases, there is no automation in place :disappointed:.

**Comment by enriqueav99** [comment]
Then we have to wait 😄. Would an automation to trigger a release fit, or are you not interested? I could try to do something

**Comment by svghadi** [comment]
It's on my todo for a while now. Need to plan it. 

> I could try to do something

Sure. I will come up with a plan and we can collaborate if you are interested.

**Comment by enriqueav99** [comment]
> It's on my todo for a while now. Need to plan it.
> 
> > I could try to do something
> 
> Sure. I will come up with a plan and we can collaborate if you are interested.

For sure! I'd be happy to help. Is there any communication channel where we can talk?

**Comment by svghadi** [comment]
You can reach out to me on CNCF slack, either in [#argo-cd-contributors](https://cloud-native.slack.com/archives/C020XM04CUW) channel or direct message.
